# 目的
本 notebook 目的為列出變數資訊以及對時間進行變數變化

# 初始環境設置

In [1]:
import sys
from pathlib import Path
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import warnings
warnings.filterwarnings('ignore')

# 設定中文字體
plt.rcParams['font.family'] = ['Microsoft JhengHei', 'DejaVu Sans']
plt.rcParams['axes.unicode_minus'] = False

# 添加專案根目錄到 Python 路徑
project_root = Path.cwd().parent
if str(project_root) not in sys.path:
    sys.path.insert(0, str(project_root))

# 匯入資料載入系統
from src.data_loader import load_data, DataLoader, ConfigLoader

# 載入配置
config = ConfigLoader.load_config(project_root / "config.yaml")
loader = DataLoader(config)

print("✅ 環境初始化完成")

2026-05-30 03:19:47,304 - src.data_loader - INFO - ✅ 設定檔已載入: c:\Users\chent\Desktop\我的資料夾\學校\大學\三下\統計方法與資料採礦\DataMiningG1FinalProject\config.yaml


✅ 環境初始化完成


In [2]:
# 載入原始資料（若本地無資料則會自動嘗試從 Kaggle 下載）
df = load_data()
print(f"原始資料形狀: {df.shape}")
display(df.head())

# 基本統計摘要
print("\n數值欄位摘要:")
display(df.describe())

print("\n所有欄位摘要:")
display(df.describe(include='all').T)

2026-05-30 03:19:47,344 - src.data_loader - INFO - ✅ 設定檔已載入: c:\Users\chent\Desktop\我的資料夾\學校\大學\三下\統計方法與資料採礦\DataMiningG1FinalProject\config.yaml
2026-05-30 03:19:47,347 - src.data_loader - INFO - ✅ 優先使用專案本地資料: c:\Users\chent\Desktop\我的資料夾\學校\大學\三下\統計方法與資料採礦\DataMiningG1FinalProject\data\raw\Buy_Now_Pay_Later_BNPL_CreditRisk_Dataset.csv
2026-05-30 03:19:47,348 - src.data_loader - INFO - 📖 讀取資料檔案: c:\Users\chent\Desktop\我的資料夾\學校\大學\三下\統計方法與資料採礦\DataMiningG1FinalProject\data\raw\Buy_Now_Pay_Later_BNPL_CreditRisk_Dataset.csv
2026-05-30 03:19:47,384 - src.data_loader - INFO - ✅ 資料已載入 | 形狀: (10345, 17)


原始資料形狀: (10345, 17)


,user_id,age,employment_type,monthly_income,credit_score,purchase_amount,product_category,bnpl_installments,repayment_delay_days,missed_payments,default_flag,app_usage_frequency,location,transaction_date,debt_to_income_ratio,risk_score,customer_segment
0,1,56,Salaried,68529.50,552,5000.00,Electronics,12,13,1,0,8.49,Australia,2023-06-10,0.072961,165.2,Medium Risk
1,2,19,Student,7247.85,300,1073.23,Fashion,12,13,1,0,3.09,USA,2024-10-07,0.148076,266.0,High Risk
2,3,20,Self-Employed,41582.26,471,5000.00,Electronics,3,19,2,0,3.33,Australia,2023-04-05,0.120244,229.6,High Risk
3,4,21,Salaried,14423.46,300,4076.83,Sports,6,18,5,1,5.86,Germany,2023-06-24,0.282653,356.0,High Risk
4,5,43,Salaried,42845.50,512,5000.00,Electronics,9,0,0,0,7.36,India,2024-10-19,0.116698,135.2,High Risk



數值欄位摘要:


,user_id,age,monthly_income,credit_score,purchase_amount,bnpl_installments,repayment_delay_days,missed_payments,default_flag,app_usage_frequency,debt_to_income_ratio,risk_score
count,10345.000000,10345.000000,10345.000000,10345.000000,10345.000000,10345.000000,10345.000000,10345.000000,10345.000000,10345.000000,10345.000000,10345.000000
mean,5173.000000,38.559884,35053.381898,448.176124,3979.402721,7.477525,8.742774,1.015950,0.390527,5.524101,0.181071,198.534094
std,2986.488601,12.131789,27084.517277,136.518332,1500.828483,3.362867,6.781849,0.996726,0.487892,2.590364,0.124887,67.541384
min,1.000000,18.000000,5000.000000,300.000000,100.000000,3.000000,0.000000,0.000000,0.000000,1.000000,0.000811,0.000000
25%,2587.000000,28.000000,12207.020000,332.000000,2990.720000,3.000000,2.000000,0.000000,0.000000,3.300000,0.080187,153.600000
50%,5173.000000,39.000000,23800.520000,403.000000,5000.000000,9.000000,9.000000,1.000000,0.000000,5.550000,0.138263,202.800000
75%,7759.000000,49.000000,55276.930000,547.000000,5000.000000,9.000000,14.000000,2.000000,1.000000,7.750000,0.268300,250.800000
max,10345.000000,59.000000,145767.120000,850.000000,5000.000000,12.000000,33.000000,7.000000,1.000000,10.000000,0.723460,398.000000



所有欄位摘要:


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
user_id,10345.0,NaN,NaN,NaN,5173.0,2986.488601,1.0,2587.0,5173.0,7759.0,10345.0
age,10345.0,NaN,NaN,NaN,38.559884,12.131789,18.0,28.0,39.0,49.0,59.0
employment_type,10345,4,Salaried,2632,NaN,NaN,NaN,NaN,NaN,NaN,NaN
monthly_income,10345.0,NaN,NaN,NaN,35053.381898,27084.517277,5000.0,12207.02,23800.52,55276.93,145767.12
credit_score,10345.0,NaN,NaN,NaN,448.176124,136.518332,300.0,332.0,403.0,547.0,850.0
purchase_amount,10345.0,NaN,NaN,NaN,3979.402721,1500.828483,100.0,2990.72,5000.0,5000.0,5000.0
product_category,10345,5,Fashion,2166,NaN,NaN,NaN,NaN,NaN,NaN,NaN
bnpl_installments,10345.0,NaN,NaN,NaN,7.477525,3.362867,3.0,3.0,9.0,9.0,12.0
repayment_delay_days,10345.0,NaN,NaN,NaN,8.742774,6.781849,0.0,2.0,9.0,14.0,33.0
missed_payments,10345.0,NaN,NaN,NaN,1.01595,0.996726,0.0,0.0,1.0,2.0,7.0


# 初始變數定義
此資料集為擬合而成，須注意

| Column | Type | Description | 處理 |
| :--- | :--- | :--- | :--- |
| `user_id` | int | Unique user identifier | 與預測無關，前處理捨棄 |
| `age` | int | User age (18–59) | no |
| `employment_type` | str | Salaried / Self-Employed / Student / Unemployed | no |
| `monthly_income` | float | Monthly income in USD | no |
| `purchase_amount` | float | BNPL transaction value in USD | no |
| `product_category` | str | Electronics, Fashion, Sports, Home, Beauty | no |
| `bnpl_installments` | int | Number of repayment installments (3, 6, 9, 12) | no |
| `repayment_delay_days` | int | Days delayed beyond due date (0–33) | no |
| `missed_payments` | int | Total missed payments (0–7) | no |
| `app_usage_frequency` | float | App opens per week | no |
| `location` | str | Country (USA, India, UK, Germany, Canada, Australia) | no |
| `transaction_date` | str | Date of purchase (YYYY-MM-DD) | 改變變數型態 |
| `debt_to_income_ratio` | float | DTI ratio (monthly debt / monthly income) | no |
| `risk_score` | float | Composite risk score (0–398) — higher = more risky | 與預測偏好意義上太相近 |
| `credit_score` | int | Standard credit score (300–850) | 與預測偏好意義上太相近 |
| `customer_segment` | str | Low Risk / Medium Risk / High Risk | 與預測偏好意義上太相近 |
| `default_flag` | int | Target: 1 = Defaulted, 0 = Paid ✅ | target |

# 時間欄位提取
從 `transaction_date` 提取年、月、日、星期幾、是否週末

In [3]:
if 'transaction_date' in df.columns:
    # 轉換為 datetime
    df['transaction_date'] = pd.to_datetime(df['transaction_date'], errors='coerce')
    
    # 提取時間特徵
    df['transaction_year'] = df['transaction_date'].dt.year
    df['transaction_month'] = df['transaction_date'].dt.month
    df['transaction_day'] = df['transaction_date'].dt.day
    df['transaction_dayofweek'] = df['transaction_date'].dt.dayofweek  # 0=Monday, 6=Sunday
    df['transaction_is_weekend'] = df['transaction_date'].dt.dayofweek.isin([5, 6]).astype(int)  #one-hot
    
    print("✅ 時間特徵工程完成")
    display(df[['transaction_date', 'transaction_year', 'transaction_month',
                'transaction_day','transaction_dayofweek', 'transaction_is_weekend']].head())
    
    df = df.drop(columns=['transaction_date'])

✅ 時間特徵工程完成


,transaction_date,transaction_year,transaction_month,transaction_day,transaction_dayofweek,transaction_is_weekend
0,2023-06-10,2023,6,10,5,1
1,2024-10-07,2024,10,7,0,0
2,2023-04-05,2023,4,5,2,0
3,2023-06-24,2023,6,24,5,1
4,2024-10-19,2024,10,19,5,1


## 改變時間類別定義
| Column | Type | Description | 處理 |
| :--- | :--- | :--- | :--- |
| `user_id` | int | Unique user identifier | 與預測無關，前處理捨棄 |
| `age` | int | User age (18–59) | no |
| `employment_type` | str | Salaried / Self-Employed / Student / Unemployed | no |
| `monthly_income` | float | Monthly income in USD | no |
| `purchase_amount` | float | BNPL transaction value in USD | no |
| `product_category` | str | Electronics, Fashion, Sports, Home, Beauty | no |
| `bnpl_installments` | int | Number of repayment installments (3, 6, 9, 12) | no |
| `repayment_delay_days` | int | Days delayed beyond due date (0–33) | no |
| `missed_payments` | int | Total missed payments (0–7) | 會產生 data leakage  |
| `app_usage_frequency` | float | App opens per week | no |
| `location` | str | Country (USA, India, UK, Germany, Canada, Australia) | no |
| `debt_to_income_ratio` | float | DTI ratio (monthly debt / monthly income) | no |
| `transaction_year` | int | Year of purchase (YYYY) | no |
| `transaction_month` | int | Month of purchase (MM) | no |
| `transaction_day` | int | day of purchase (DD) | no |
| `transaction_dayofweek` | int | Day of week of purchase | no |
| `transaction_is_weekend` | bool | Date of purchase is weekend | no |
| `risk_score` | float | Composite risk score (0–398) — higher = more risky | 與預測偏好意義上太相近 |
| `credit_score` | int | Standard credit score (300–850) | 與預測偏好意義上太相近 |
| `customer_segment` | str | Low Risk / Medium Risk / High Risk | 與預測偏好意義上太相近 |
| `default_flag` | int | Target: 1 = Defaulted, 0 = Paid ✅ | target |

## 儲存時間處理結果 

In [4]:
# 儲存通用預處理資料
loader.save_interim_data(df, "01_data_understanding.csv")

print("✅ 所有預處理資料已儲存至 data/interim/")

2026-05-30 03:19:47,621 - src.data_loader - INFO - ✅ 中間資料已儲存 (interim): c:\Users\chent\Desktop\我的資料夾\學校\大學\三下\統計方法與資料採礦\DataMiningG1FinalProject\data\interim\01_data_understanding.csv


✅ 所有預處理資料已儲存至 data/interim/
